ocaml/xapi/quicktest_cbt.ml

```diff
@@ -0,0 +1,101 @@
+  let open Client in
+
+  (* Helper for test failure due to unexpected error *)
+  let report_failure error test_name test =
+    failed test (Printf.sprintf "%s failed: %s" test_name
+                   (ExnHelper.string_of_exn error)) in
+
+
+  let test_assert ~test op ~msg =
+    try assert op with assert_failure -> failed test msg in
+
+  (* Overall test suite runs smaller unit tests *)
+  let cbt_test = make_test "Testing changed block tracking" 2 in
+  try
+    start cbt_test;
+
+    (* generate list of SRs attached to a host *)
+    let list_of_attached_srs =
+      let is_attached sr =
+        let pbds = SR.get_PBDs ~session_id ~rpc:!rpc ~self:sr in
+        List.fold_left (||) false
+          (List.map (fun pbd ->
+               PBD.get_currently_attached ~session_id ~rpc:!rpc ~self:pbd
+             ) pbds) in
+      List.filter is_attached (SR.get_all ~session_id ~rpc:!rpc) in
+
+    (* find lvm SR (that passes data_destroy) to attach new VDI *)
+    let sR = List.filter
+        (fun sr ->
+           ((SR.get_type ~session_id ~rpc:!rpc ~self:sr) = "lvm")
+           && (not (SR.get_is_tools_sr ~session_id ~rpc:!rpc ~self:sr))
+        ) list_of_attached_srs |> List.hd in
+
+    let sR_operations =
+      SR.get_allowed_operations ~session_id ~rpc:!rpc ~self:sR in
+
+    let vDI = VDI.create ~session_id ~rpc:!rpc ~name_label:"qt-cbt"
+        ~name_description:"VDI for CBT quicktest" ~_type:`user ~sR
+        ~sharable:false ~other_config:[] ~read_only:false ~sm_config:[]
+        ~virtual_size:(2L ** mib) ~xenstore_data:[] ~tags:[] in
+
+    (* Test enable/disable CBT, test cbt_enabled:false for new VDI *)
+    let enable_disable_cbt_test ~vDI =
+      let enable_cbt_test = make_test "Testing VDI.enable/disable_CBT" 4 in
+      let get_cbt_status vDI = VDI.get_cbt_enabled ~session_id ~rpc:!rpc ~self:vDI in
+      let test = enable_cbt_test in
+      try
+        start enable_cbt_test;
+        test_assert ~test
+          (not (VDI.get_cbt_enabled ~session_id ~rpc:!rpc ~self:vDI))
+          ~msg:"VDI.cbt_enabled field should be set to false for new VDIs, but wasn't ";
+        VDI.enable_cbt ~session_id ~rpc:!rpc ~self:vDI;
+        test_assert ~test
+          (get_cbt_status vDI)
+          ~msg:"VDI.enable_cbt failed";
+        VDI.disable_cbt ~session_id ~rpc:!rpc ~self:vDI;
+        test_assert ~test
+          (not (get_cbt_status vDI)) (* disable_cbt fails *)
+          ~msg:"VDI.disable_CBT failed";
+        success enable_cbt_test
+      with e -> report_failure e "enable/disable CBT" enable_cbt_test in
+
+    (* Call unit tests if and only if SR supports required operations *)
+    let run_tests ~vDI ~sR_operations =
+      List.iter (fun (test, required_operations) ->
+          if List.for_all
+              (fun x -> List.mem x sR_operations)
+              required_operations
+          then test ~vDI
+          else ignore ())
+        [ enable_disable_cbt_test , [`vdi_enable_cbt ; `vdi_disable_cbt]
```

Comment:
Related to my above comment, I think we could make a small change here: loop through all the SRs, instead of picking the first LVM SR, and then getting their allowed operations, and then for each SR we loop through each test and run that test if that SR supports the operations needed by that test. This way, we will also check that we correctly defined the allowed_operations field of the SR, since if we say an SR supports `vdi_enable_cbt`, when in fact it doesn't, our test will fail.

Reply:
So, iterate the test suite over each SR? I think that might add a significant amount to quicktest run time.

Reply:
I think that's nice because we do not make any assumptions about the environment, so the quicktest will work for all setups, which is really nice :). And this is what the storage quicktests do. In practice, when we run the quicktests, I think there will be only one SR which supports the operations we need, and that SR will be an LVM SR probably, so I think it shouldn't slow down the quicktests that we run as part of our build. But I think it's still useful to have this logic in case somebody wants to run the tests for various SRs, they can then just create the SRs and run the quicktests.